File: static_site/src/utils/programGenerator.js
/**
 * Program Generator for 5/3/1 Krypteia
 * Generates a 13-week training program based on user's training maxes
 */

/**
 * Round weight according to rounding rules
 */
function roundWeight(weight, units) {
  const increment = units === 'lb' ? 5 : 2.5;
  return Math.round(weight / increment) * increment;
}

/**
 * Calculate weight for a set based on TM and percentage
 */
function calculateSetWeight(tm, pctTM, units) {
  const rawWeight = tm * pctTM;
  return roundWeight(rawWeight, units);
}

/**
 * Generate sets for a given scheme and training max
 */
function generateSets(scheme, tm, units) {
  return scheme.workSets.map(set => ({
    weight: calculateSetWeight(tm, set.pctTM, units),
    targetReps: set.reps,
    pctTM: set.pctTM,
  }));
}

/**
 * Get the set scheme for a given week in the cycle
 */
function getSetSchemeForWeek(template, phase, weekInCycle) {
  if (phase.mainLiftScheme) {
    return template.setSchemes[phase.mainLiftScheme];
  }
  
  if (phase.mainLiftSchemeByWeekInCycle) {
    const weekInPhase = ((weekInCycle - 1) % 3) + 1;
    const schemeName = phase.mainLiftSchemeByWeekInCycle[weekInPhase.toString()];
    return template.setSchemes[schemeName];
  }
  
  return null;
}

/**
 * Generate a single session for a given week
 */
function generateSession(template, sessionTemplate, phase, weekInCycle, trainingMaxes, units) {
  const liftId = sessionTemplate.mainLiftId;
  const tm = trainingMaxes[liftId];
  
  const setScheme = getSetSchemeForWeek(template, phase, weekInCycle);
  
  if (!setScheme) {
    // Test week or special week - return minimal structure
    return {
      sessionId: sessionTemplate.sessionId,
      label: sessionTemplate.label,
      mainLiftId: liftId,
      phase: phase.phaseId,
      isTestWeek: phase.phaseId === 'TEST',
      isReset: phase.phaseId === 'RESET',
    };
  }
  
  const mainSets = generateSets(setScheme, tm, units);
  
  // Generate supplemental (FSL) if enabled
  let supplemental = null;
  if (phase.rules?.supplementalUsesFSL && phase.rules?.supplementalEnabled !== false) {
    // FSL uses the FIRST set weight of the current week's scheme
    const firstSetPercent = setScheme.workSets[0].pctTM;
    const fslWeight = calculateSetWeight(tm, firstSetPercent, units);
    const fslTemplate = sessionTemplate.supplemental?.find(s => s.type === 'fsl_main_lift');
    if (fslTemplate) {
      supplemental = {
        type: 'fsl_main_lift',
        label: fslTemplate.label,
        sets: fslTemplate.sets,
        repsRange: fslTemplate.repsRange,
        weight: fslWeight,
      };
    }
  }
  
  return {
    sessionId: sessionTemplate.sessionId,
    label: sessionTemplate.label,
    mainLiftId: liftId,
    phase: phase.phaseId,
    setScheme: setScheme.label,
    mainSets,
    supplemental,
    assistanceSlots: sessionTemplate.assistanceSlots,
    circuit: {
      ...sessionTemplate.circuit,
      rounds: phase.rules?.circuitRounds || sessionTemplate.circuit?.rounds || 5,
    },
    conditioningSlots: sessionTemplate.conditioningSlots,
    mobilitySlots: sessionTemplate.mobilitySlots,
    rules: phase.rules,
  };
}

/**
 * Generate a single week of training
 */
function generateWeek(template, weekNumber, trainingMaxes, units) {
  // Find which phase this week belongs to
  const phase = template.macrocycle.phases.find(p => p.weeks.includes(weekNumber));
  
  if (!phase) {
    throw new Error(`Week ${weekNumber} not found in any phase`);
  }
  
  const sessions = template.macrocycle.weeklySessions.map(sessionSlot => {
    const sessionTemplate = template.sessionTemplates[sessionSlot.sessionTemplateRef];
    return generateSession(template, sessionTemplate, phase, weekNumber, trainingMaxes, units);
  });
  
  return {
    weekNumber,
    phase: phase.phaseId,
    phaseLabel: phase.label,
    sessions,
  };
}

/**
 * Generate the full 13-week program
 */
export function generateProgram(template, trainingMaxes, units) {
  const weeks = [];
  
  for (let weekNum = 1; weekNum <= template.macrocycle.cycleLengthWeeks; weekNum++) {
    weeks.push(generateWeek(template, weekNum, trainingMaxes, units));
  }
  
  return {
    programId: template.programId,
    programName: template.programName,
    version: template.version,
    generatedAt: new Date().toISOString(),
    trainingMaxes,
    units,
    weeks,
  };
}

/**
 * Calculate training maxes from 1RMs
 */
export function calculateTrainingMaxes(oneRepMaxes, tmPercent = 0.85, units = 'lb') {
  return {
    squat: roundWeight(oneRepMaxes.squat * tmPercent, units),
    bench: roundWeight(oneRepMaxes.bench * tmPercent, units),
    deadlift: roundWeight(oneRepMaxes.deadlift * tmPercent, units),
    ohp: roundWeight(oneRepMaxes.ohp * tmPercent, units),
  };
}

/**
 * Load the plan template from public folder
 */
export async function loadPlanTemplate() {
  const response = await fetch('/plan.template.json');
  if (!response.ok) {
    throw new Error('Failed to load plan template');
  }
  return response.json();
}

File: mobile/strykr/strykr/ProgramGenerator.swift
import Foundation

struct ProgramTemplate: Codable {
    let programName: String
    let macrocycle: Macrocycle
    let setSchemes: [String: SetScheme]
    let sessionTemplates: [String: SessionTemplate]
}

struct Macrocycle: Codable {
    let cycleLengthWeeks: Int
    let phases: [Phase]
}

struct Phase: Codable {
    let phaseId: String
    let label: String
    let weeks: [Int]
    let mainLiftScheme: String?
    let mainLiftSchemeByWeekInCycle: [String: String]?
    let rules: PhaseRules
}

struct PhaseRules: Codable {
    let supplementalUsesFSL: Bool?
    let supplementalEnabled: Bool?
    let circuitRounds: Int?
}

struct SetScheme: Codable {
    let label: String
    let workSets: [WorkSet]
}

struct WorkSet: Codable {
    let pctTM: Double
    let reps: Int
}

struct SessionTemplate: Codable {
    let sessionId: String
    let label: String
    let mainLiftId: String
    let supplemental: [SupplementalTemplate]
    let assistanceSlots: [AssistanceSlot]
    let circuit: CircuitTemplate
}

struct SupplementalTemplate: Codable {
    let type: String
    let label: String
    let sets: Int
    let repsRange: [Int]
}

struct AssistanceSlot: Codable {
    let slotId: String
    let minReps: Int
    let maxReps: Int
}

struct CircuitTemplate: Codable {
    let enabled: Bool
    let rounds: Int
    let style: String
}


struct GeneratedProgram {
    let weeks: [ProgramWeek]
    let startDate: Date
}

struct ProgramWeek {
    let weekNumber: Int
    let sessions: [ProgramSession]
}

struct ProgramSession {
    let sessionId: String
    let sessionName: String
    let date: Date
    let mainLift: MainLiftDetails
    let circuit: CircuitDetails
}

struct MainLiftDetails {
    let liftId: String
    let liftName: String
    let sets: [SetDetails]
}

struct CircuitDetails {
    let rounds: Int
    let exercises: [CircuitExercise]
}

struct CircuitExercise {
    let exerciseId: String
    let exerciseName: String
    let sets: [SetDetails]
}

struct SetDetails {
    let weight: Double
    let targetReps: Int
    let pctTM: Double
}

class ProgramGenerator {
    static func generateProgram(
        template: ProgramTemplate,
        strengthData: StrengthData,
        profile: Profile,
        exercises: [Exercise]
    ) -> GeneratedProgram {
        let startDate = calculateStartDate(preferredStartDay: profile.preferredStartDay)
        var weeks: [ProgramWeek] = []
        
        for weekNum in 1...template.macrocycle.cycleLengthWeeks {
            guard let phase = template.macrocycle.phases.first(where: { $0.weeks.contains(weekNum) }) else {
                continue
            }
            
            // Determine main lift scheme for this week
            let mainSchemeName: String
            if let schemeByWeek = phase.mainLiftSchemeByWeekInCycle {
                let weekInPhase = weekNum - (phase.weeks.first ?? 1) + 1
                mainSchemeName = schemeByWeek["\(weekInPhase)"] ?? phase.mainLiftScheme ?? "fives_pro_week_1"
            } else {
                mainSchemeName = phase.mainLiftScheme ?? "fives_pro_week_1"
            }
            
            guard let mainScheme = template.setSchemes[mainSchemeName] else {
                continue
            }
            
            var sessions: [ProgramSession] = []
            let sessionKeys = ["SQUAT_DAY", "BENCH_DAY", "DEADLIFT_DAY", "OHP_DAY"]
            
            for (dayOffset, sessionKey) in sessionKeys.enumerated() {
                guard let sessionTemplate = template.sessionTemplates[sessionKey] else {
                    continue
                }
                
                let sessionDate = startDate.addingTimeInterval(Double((weekNum - 1) * 7 + dayOffset) * 86400)
                
                let mainLift = generateMainLift(
                    liftId: sessionTemplate.mainLiftId,
                    scheme: mainScheme,
                    strengthData: strengthData
                )
                
                let circuit = generateCircuit(
                    sessionTemplate: sessionTemplate,
                    mainLiftId: sessionTemplate.mainLiftId,
                    mainScheme: mainScheme,
                    strengthData: strengthData,
                    exercises: exercises,
                    phaseRules: phase.rules
                )
                
                sessions.append(ProgramSession(
                    sessionId: sessionTemplate.sessionId,
                    sessionName: sessionTemplate.label,
                    date: sessionDate,
                    mainLift: mainLift,
                    circuit: circuit
                ))
            }
            
            weeks.append(ProgramWeek(weekNumber: weekNum, sessions: sessions))
        }
        
        return GeneratedProgram(weeks: weeks, startDate: startDate)
    }
    
    private static func calculateStartDate(preferredStartDay: String) -> Date {
        let calendar = Calendar.current
        let today = Date()
        let todayWeekday = calendar.component(.weekday, from: today)
        
        let targetWeekday: Int = {
            switch preferredStartDay.lowercased() {
            case "sun": return 1
            case "mon": return 2
            case "tue": return 3
            case "wed": return 4
            case "thu": return 5
            case "fri": return 6
            case "sat": return 7
            default: return 2
            }
        }()
        
        var daysToAdd = targetWeekday - todayWeekday
        if daysToAdd <= 0 {
            daysToAdd += 7
        }
        
        return calendar.date(byAdding: .day, value: daysToAdd, to: today)!
    }
    
    private static func generateMainLift(
        liftId: String,
        scheme: SetScheme,
        strengthData: StrengthData
    ) -> MainLiftDetails {
        let tm = getTrainingMax(liftId: liftId, strengthData: strengthData)
        let liftName = liftId.capitalized
        
        let sets = scheme.workSets.map { workSet in
            let weight = round(tm * workSet.pctTM / 5) * 5
            return SetDetails(weight: weight, targetReps: workSet.reps, pctTM: workSet.pctTM)
        }
        
        return MainLiftDetails(liftId: liftId, liftName: liftName, sets: sets)
    }
    
    private static func generateCircuit(
        sessionTemplate: SessionTemplate,
        mainLiftId: String,
        mainScheme: SetScheme,
        strengthData: StrengthData,
        exercises: [Exercise],
        phaseRules: PhaseRules
    ) -> CircuitDetails {
        var circuitExercises: [CircuitExercise] = []
        let rounds = phaseRules.circuitRounds ?? sessionTemplate.circuit.rounds
        
        // Add FSL (main lift)
        if let fslSupplemental = sessionTemplate.supplemental.first(where: { $0.type == "fsl_main_lift" }) {
            let tm = getTrainingMax(liftId: mainLiftId, strengthData: strengthData)
            let firstSetPct = mainScheme.workSets.first?.pctTM ?? 0.65
            let weight = round(tm * firstSetPct / 5) * 5
            
            let sets = (0..<fslSupplemental.sets).map { _ in
                SetDetails(weight: weight, targetReps: fslSupplemental.repsRange[0], pctTM: firstSetPct)
            }
            
            circuitExercises.append(CircuitExercise(
                exerciseId: mainLiftId,
                exerciseName: mainLiftId.capitalized,
                sets: sets
            ))
        }
        
        // Add assistance exercises
        for slot in sessionTemplate.assistanceSlots {
            if let exercise = selectRandomExercise(forTag: slot.slotId, from: exercises) {
                let sets = (0..<5).map { _ in
                    SetDetails(weight: 0, targetReps: slot.minReps, pctTM: 0)
                }
                
                circuitExercises.append(CircuitExercise(
                    exerciseId: exercise.id,
                    exerciseName: exercise.name,
                    sets: sets
                ))
            }
        }
        
        return CircuitDetails(rounds: rounds, exercises: circuitExercises)
    }
    
    private static func getTrainingMax(liftId: String, strengthData: StrengthData) -> Double {
        switch liftId {
        case "squat": return strengthData.trainingMaxes.squat
        case "bench": return strengthData.trainingMaxes.bench
        case "deadlift": return strengthData.trainingMaxes.deadlift
        case "ohp": return strengthData.trainingMaxes.ohp
        default: return 0
        }
    }
    
    private static func selectRandomExercise(forTag tag: String, from exercises: [Exercise]) -> Exercise? {
        let matching = exercises.filter { $0.slotTags.contains(tag) }
        return matching.randomElement()
    }
}

